optuna/storages/_rdb/storage.py

```diff
@@ -463,7 +463,45 @@ def _create_new_trial(
         """
 
         session = self.scoped_session()
+        try:
+            # Lock all trials belonging to this study. This might lead to a deadlock
+            # (`OperationalError`) in which case we will retry.
+            session.query(models.TrialModel).filter(
+                models.TrialModel.study_id == study_id
+            ).with_for_update().all()
```

Comment:
How about locking a entry in the `studies` table instead of locking the whole trials (though I haven't checked that it improves the latency)?

Reply:
Ah, that might scale better with the number of trials. I'll try your suggestion.

Reply:
Looks good.

| #params | #trials | time(sec) |
| ------- | ------- | --------- |
| 1 | 1 | 0.02 |
| 1 | 10 | 0.11 |
| 1 | 100 | 1.26 |
| 1 | 1000 | 17.41 |
| 2 | 1 | 0.02 |
| 2 | 10 | 0.12 |
| 2 | 100 | 1.32 |
| 2 | 1000 | 20.22 |
| 4 | 1 | 0.02 |
| 4 | 10 | 0.13 |
| 4 | 100 | 1.61 |
| 4 | 1000 | 25.82 |
| 8 | 1 | 0.04 |
| 8 | 10 | 0.16 |
| 8 | 100 | 2.07 |
| 8 | 1000 | 35.07 |
| 16 | 1 | 0.06 |
| 16 | 10 | 0.20 |
| 16 | 100 | 3.02 |
| 16 | 1000 | 55.23 |
| 32 | 1 | 0.12 |
| 32 | 10 | 0.30 |
| 32 | 100 | 4.76 |
| 32 | 1000 | 90.28 |